src/mappings/mappingHandlers.ts

```diff
@@ -284,3 +286,83 @@ export async function handleDelegatorWithdrawRewardEvent(event: CosmosEvent): Pr
   claim.denom = denom;
   await claim.save();
 }
+
+export async function handleNativeBalanceDecrement(event: CosmosEvent): Promise<void> {
+  logger.info(`[handleNativeBalanceDecrement] (tx ${event.tx.hash}): indexing event ${event.idx + 1} / ${event.tx.tx.events.length}`)
+  logger.debug(`[handleNativeBalanceDecrement] (event.event): ${JSON.stringify(event.event, null, 2)}`)
+  logger.debug(`[handleNativeBalanceDecrement] (event.log): ${JSON.stringify(event.log, null, 2)}`)
+
+  // sample event.event.attributes:
+  // [
+  //   {"key":"spender","value":"fetch1jv65s3grqf6v6jl3dp4t6c9t9rk99cd85zdctg"},
+  //   {"key":"amount","value":"75462013217046121atestfet"},
+  //   {"key":"spender","value":"fetch1wurz7uwmvchhc8x0yztc7220hxs9jxdjdsrqmn"},
+  //   {"key":"amount","value":"100atestfet"}
+  // ]
+  let spendEvents = [];
+  event.event.attributes.map((e, i) => {
+    if (e.key !== "spender") {
+      return
+    }
+    const spender = e.value;
+    const amountStr = event.event.attributes[i+1].value;
+
+    const coin = parseCoins(amountStr)[0];
+    const amount = BigInt(0) - BigInt(coin.amount); // save a negative amount for a "spend" event
+    spendEvents.push({spender: spender, amount: amount, denom: coin.denom})
+  });
+  
+  for (const i in spendEvents) {
+    const spendEvent = spendEvents[i];
+    await saveNativeBalanceEvent(`${messageId(event)}-spend-${i}`, spendEvent.spender, spendEvent.amount, spendEvent.denom, event);
+  }
+}
+
+export async function handleNativeBalanceIncrement(event: CosmosEvent): Promise<void> {
+  logger.info(`[handleNativeBalanceIncrement] (tx ${event.tx.hash}): indexing event ${event.idx + 1} / ${event.tx.tx.events.length}`)
+  logger.debug(`[handleNativeBalanceIncrement] (event.event): ${JSON.stringify(event.event, null, 2)}`)
+  logger.debug(`[handleNativeBalanceIncrement] (event.log): ${JSON.stringify(event.log, null, 2)}`)
+
+  // sample event.event.attributes:
+  // [
+  //   {"key":"receiver","value":"fetch1jv65s3grqf6v6jl3dp4t6c9t9rk99cd85zdctg"},
+  //   {"key":"amount","value":"75462013217046121atestfet"},
+  //   {"key":"receiver","value":"fetch1wurz7uwmvchhc8x0yztc7220hxs9jxdjdsrqmn"},
+  //   {"key":"amount","value":"100atestfet"}
+  // ]
+  let receiveEvents = [];
+  event.event.attributes.map((e, i) => {
+    if (e.key !== "receiver") {
+      return
+    }
+    const receiver = e.value;
+    const amountStr = event.event.attributes[i+1].value;
+
+    const coin = parseCoins(amountStr)[0];
+    const amount = BigInt(coin.amount);
+    receiveEvents.push({receiver: receiver, amount: amount, denom: coin.denom})
+  });
+  
+  for (const i in receiveEvents) {
+    const receiveEvent = receiveEvents[i];
```

Comment:
I think this can be simplified with [`for...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of), which is for looping over iterables. Whereas, [`for...in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) is for looping over "enumerable properties of an object".

```suggestion
  for (const receiveEvent of receiveEvents) {
```

Reply:
yea made that first, but the `i` is actually needed to generate the row ID in `${messageId(event)}-receive-${i}` below

Reply:
Ah, I missed that detail, thanks. Perhaps the `Object.entries` approach then?

Reply:
Yep so I just pushed a rework using `Object.entries` everywhere. Not sure I like it better as:
- we give an array to Object.entries(), which feels weird but well, javascript maybe.
- Object.entries somehow cast the array index to string (probably because as the name suggest, it expect an actual object rather than an array). So the events attributes become: `[["0",{"key":"receiver","value":"fetch147s4ljcxuqpg0tg3kkwspthjau0mjzknzpm53j"}],["1",{"key":"amount","value":"10000000000000000000atestfet"}]]`. And javascript is quite happy to do "1" + 1, but yield "11" rather than 2 - so parseInt required for the array access few line below, and 30 minutes wasted trying to figure what broke.

Reply:
Oh boy, JavaScript. My apologies for for accidentally turning on the wat-signal.